clients/src/main/java/org/apache/kafka/common/security/JaasContext.java

```diff
@@ -0,0 +1,185 @@
+     * listenerName is not defined for contextType SERVER of if listenerName is defined for contextType CLIENT.
+     */
+    public static JaasContext load(JaasContext.Type contextType, ListenerName listenerName,
+                                   Map<String, ?> configs) {
+        String listenerContextName;
+        String globalContextName;
+        switch (contextType) {
+            case CLIENT:
+                if (listenerName != null)
+                    throw new IllegalArgumentException("listenerName should be null for CLIENT");
+                globalContextName = GLOBAL_CONTEXT_NAME_CLIENT;
+                listenerContextName = null;
+                break;
+            case SERVER:
+                if (listenerName == null)
+                    throw new IllegalArgumentException("listenerName should not be null for SERVER");
+                globalContextName = GLOBAL_CONTEXT_NAME_SERVER;
+                listenerContextName = listenerName.value().toLowerCase(Locale.ROOT) + "." + GLOBAL_CONTEXT_NAME_SERVER;
+                break;
+            default:
+                throw new IllegalArgumentException("Unexpected context type " + contextType);
+        }
+        return load(contextType, listenerContextName, globalContextName, configs);
+    }
+
+    static JaasContext load(JaasContext.Type contextType, String listenerContextName,
+                            String globalContextName, Map<String, ?> configs) {
+        Password jaasConfigArgs = (Password) configs.get(SaslConfigs.SASL_JAAS_CONFIG);
+        if (jaasConfigArgs != null) {
+            if (contextType == JaasContext.Type.SERVER)
+                throw new IllegalArgumentException("JAAS config property not supported for server");
+            else {
+                JaasConfig jaasConfig = new JaasConfig(globalContextName, jaasConfigArgs.value());
+                AppConfigurationEntry[] clientModules = jaasConfig.getAppConfigurationEntry(globalContextName);
+                int numModules = clientModules == null ? 0 : clientModules.length;
+                if (numModules != 1)
+                    throw new IllegalArgumentException("JAAS config property contains " + numModules + " login modules, should be 1 module");
+                return new JaasContext(globalContextName, contextType, jaasConfig);
+            }
+        } else
+            return defaultContext(contextType, listenerContextName, globalContextName);
+    }
+
+    private static JaasContext defaultContext(JaasContext.Type contextType, String listenerContextName,
+                                              String globalContextName) {
+        String jaasConfigFile = System.getProperty(JaasUtils.JAVA_LOGIN_CONFIG_PARAM);
+        if (jaasConfigFile == null) {
+            LOG.debug("System property '" + JaasUtils.JAVA_LOGIN_CONFIG_PARAM + "' and Kafka SASL property '" +
+                    SaslConfigs.SASL_JAAS_CONFIG + "' are not set, using default JAAS configuration.");
```

Comment:
Since SaslConfigs.SASL_JAAS_CONFIG can only be set on the client, perhaps we can give different logging depending on the context Type?